File: macos/DSHLauncher/Services/DSHEventMonitor.swift
import Foundation

struct DSHNotification: Sendable {
    let title: String
    let message: String
    let isCompletion: Bool
}

actor DSHEventMonitor {
    typealias StatusHandler = @MainActor @Sendable (DSHStatusSnapshot) -> Void
    typealias NotificationHandler = @MainActor @Sendable (DSHNotification) -> Void

    private var monitorTask: Task<Void, Never>?
    private var runningSessions = Set<String>()
    private var runningJobs = Set<String>()
    private var pendingQuestions = Set<String>()
    private var pendingApprovals = Set<String>()
    private var notifiedAttention = Set<String>()
    private var titles: [String: String] = [:]
    private let logger: AppLogger
    private let statusHandler: StatusHandler
    private let notificationHandler: NotificationHandler

    init(logger: AppLogger, statusHandler: @escaping StatusHandler, notificationHandler: @escaping NotificationHandler) {
        self.logger = logger
        self.statusHandler = statusHandler
        self.notificationHandler = notificationHandler
    }

    func start(port: Int) {
        monitorTask?.cancel()
        monitorTask = Task { [weak self] in await self?.monitorLoop(port: port) }
    }

    func stop() {
        monitorTask?.cancel()
        monitorTask = nil
        clear()
        publish(ready: false)
    }

    private func monitorLoop(port: Int) async {
        var delay = 2.0
        while !Task.isCancelled {
            do {
                guard await loadBaseline(port: port) else {
                    clear(); publish(ready: false)
                    try await Task.sleep(for: .seconds(delay))
                    delay = min(15, delay * 1.6)
                    continue
                }
                delay = 2
                publish(ready: true)
                try await withThrowingTaskGroup(of: Void.self) { group in
                    group.addTask { try await self.receive(path: "/api/events.host", port: port) }
                    group.addTask { try await self.receive(path: "/api/events.mux", port: port) }
                    _ = try await group.next()
                    group.cancelAll()
                }
            } catch is CancellationError { return }
            catch { await logger.warning("DSH event connection failed: \(error.localizedDescription)") }
            try? await Task.sleep(for: .seconds(delay))
        }
    }

    private func receive(path: String, port: Int) async throws {
        var request = URLRequest(url: URL(string: "ws://127.0.0.1:\(port)\(path)")!)
        request.setValue("http://127.0.0.1:\(port)", forHTTPHeaderField: "Origin")
        let socket = URLSession.shared.webSocketTask(with: request)
        socket.resume()
        await logger.info("Connected to \(path)")
        defer { socket.cancel(with: .goingAway, reason: nil) }
        while !Task.isCancelled {
            let message = try await socket.receive()
            switch message {
            case .string(let text): processFrame(Data(text.utf8))
            case .data(let data): processFrame(data)
            @unknown default: break
            }
        }
    }

    private func processFrame(_ data: Data) {
        guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let method = root["method"] as? String,
              let payload = root["payload"] as? [String: Any] else { return }
        let sessionID = payload["sessionId"] as? String ?? "unknown"
        switch method {
        case "host/session-status": set(&runningSessions, key: sessionID, enabled: payload["running"] as? Bool == true)
        case "question/requested":
            let key = "question:\(sessionID)"; pendingQuestions.insert(key)
            if notifiedAttention.insert(key).inserted {
                let questions = payload["questions"] as? [[String: Any]]
                let message = questions?.first?["header"] as? String ?? "打开 DSH 回答待处理问题。"
                sendNotification(DSHNotification(title: "DSH 需要你的回答", message: message, isCompletion: false))
            }
        case "question/resolved":
            pendingQuestions = pendingQuestions.filter { !$0.hasPrefix("question:\(sessionID)") }
            notifiedAttention.remove("question:\(sessionID)")
        case "approval/requested":
            let id = payload["approvalId"] as? String ?? sessionID; let key = "approval:\(id)"
            pendingApprovals.insert(key)
            if notifiedAttention.insert(key).inserted {
                sendNotification(DSHNotification(
                    title: "DSH 请求权限审批",
                    message: payload["toolName"] as? String ?? "打开 DSH 查看审批请求。",
                    isCompletion: false
                ))
            }
        case "approval/resolved":
            let id = payload["approvalId"] as? String ?? sessionID
            pendingApprovals.remove("approval:\(id)"); notifiedAttention.remove("approval:\(id)")
        case "session/jobs": processJobs(sessionID: sessionID, payload: payload)
        case "session/projection":
            if payload["key"] as? String == "title", let title = payload["value"] as? String { titles[sessionID] = title }
        case "session/event": processSessionEvent(sessionID: sessionID, payload: payload)
        default: break
        }
        publish(ready: true)
    }

    private func processJobs(sessionID: String, payload: [String: Any]) {
        runningJobs = runningJobs.filter { !$0.hasPrefix(sessionID + ":") }
        for job in payload["jobs"] as? [[String: Any]] ?? [] {
            if let status = job["status"] as? String, status == "running" || status == "stopping" {
                runningJobs.insert(sessionID + ":" + (job["id"] as? String ?? UUID().uuidString))
            }
        }
    }

    private func processSessionEvent(sessionID: String, payload: [String: Any]) {
        guard let event = payload["event"] as? [String: Any], let type = event["type"] as? String else { return }
        if type == "turn/start" { runningSessions.insert(sessionID) }
        if type == "turn/end" {
            runningSessions.remove(sessionID)
            let title = titles[sessionID] ?? "DSH 对话"
            Task {
                try? await Task.sleep(for: .seconds(3))
                await notificationHandler(DSHNotification(title: "对话完成", message: title, isCompletion: true))
            }
        }
    }

    private func loadBaseline(port: Int) async -> Bool {
        guard let url = URL(string: "http://127.0.0.1:\(port)/api/session.list") else { return false }
        var request = URLRequest(url: url); request.httpMethod = "POST"; request.timeoutInterval = 4
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONSerialization.data(withJSONObject: [
            "type": "client-request", "rpcId": UUID().uuidString, "method": "session.list", "payload": [:]
        ])
        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard (response as? HTTPURLResponse)?.statusCode == 200,
                  let root = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let result = root["result"] as? [String: Any], result["ok"] as? Bool == true else { return false }
            runningSessions.removeAll()
            if let value = result["value"] as? [String: Any], let items = value["items"] as? [[String: Any]] {
                for item in items {
                    guard let id = item["sessionId"] as? String else { continue }
                    if item["running"] as? Bool == true { runningSessions.insert(id) }
                }
            }
            return true
        } catch { return false }
    }

    private func publish(ready: Bool) {
        guard ready else {
            let handler = statusHandler
            Task { @MainActor in handler(.stopped) }
            return
        }
        let attention = pendingQuestions.count + pendingApprovals.count
        let busy = runningSessions.count + runningJobs.count
        let state: DSHActivityState = attention > 0 ? .attention : busy > 0 ? .busy : .idle
        let snapshot = DSHStatusSnapshot(
            state: state,
            runningSessions: runningSessions.count,
            runningJobs: runningJobs.count,
            pendingQuestions: pendingQuestions.count,
            pendingApprovals: pendingApprovals.count,
            summary: attention > 0 ? "DSH 等待处理" : busy > 0 ? "DSH 正在运行任务" : "DSH 已运行，当前空闲"
        )
        let handler = statusHandler
        Task { @MainActor in handler(snapshot) }
    }

    private func clear() {
        runningSessions.removeAll(); runningJobs.removeAll(); pendingQuestions.removeAll(); pendingApprovals.removeAll(); titles.removeAll()
    }

    private func set(_ set: inout Set<String>, key: String, enabled: Bool) {
        if enabled { set.insert(key) } else { set.remove(key) }
    }

    private func sendNotification(_ notification: DSHNotification) {
        let handler = notificationHandler
        Task { @MainActor in handler(notification) }
    }
}

File: src/DshLauncher/Services/DshEventMonitor.cs
using System.Net.Http.Json;
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;
using DshLauncher.Infrastructure;
using DshLauncher.Models;

namespace DshLauncher.Services;

internal sealed class DshEventMonitor : IAsyncDisposable
{
    private readonly AppLogger _log;
    private readonly HttpClient _http = new() { Timeout = TimeSpan.FromSeconds(8) };
    private readonly object _sync = new();
    private readonly HashSet<string> _runningSessions = new(StringComparer.Ordinal);
    private readonly HashSet<string> _runningJobs = new(StringComparer.Ordinal);
    private readonly HashSet<string> _pendingQuestions = new(StringComparer.Ordinal);
    private readonly HashSet<string> _pendingApprovals = new(StringComparer.Ordinal);
    private readonly HashSet<string> _notifiedAttention = new(StringComparer.Ordinal);
    private readonly Dictionary<string, string> _titles = new(StringComparer.Ordinal);
    private CancellationTokenSource? _lifetime;
    private Task? _loop;
    private System.Threading.Timer? _completionTimer;
    private DshNotification? _pendingCompletion;
    private int _port;

    public DshEventMonitor(AppLogger log)
    {
        _log = log;
    }

    public event EventHandler<DshStatusSnapshot>? StatusChanged;
    public event EventHandler<DshNotification>? NotificationRequested;
    public event EventHandler<Uri>? SocketConnected;

    public void Start(int port)
    {
        if (_lifetime is not null && _port == port)
        {
            return;
        }

        Stop();
        _port = port;
        _lifetime = new CancellationTokenSource();
        _loop = MonitorLoopAsync(port, _lifetime.Token);
    }

    public void Stop()
    {
        if (_lifetime is null)
        {
            return;
        }

        _lifetime.Cancel();
        _lifetime.Dispose();
        _lifetime = null;
        _loop = null;
        ClearTransientState();
        PublishStatus(false);
    }

    private async Task MonitorLoopAsync(int port, CancellationToken cancellationToken)
    {
        var delay = TimeSpan.FromSeconds(2);
        while (!cancellationToken.IsCancellationRequested)
        {
            try
            {
                if (!await IsReadyAsync(port, cancellationToken))
                {
                    ClearTransientState();
                    PublishStatus(false);
                    await Task.Delay(delay, cancellationToken);
                    delay = TimeSpan.FromSeconds(Math.Min(15, delay.TotalSeconds * 1.6));
                    continue;
                }

                delay = TimeSpan.FromSeconds(2);
                ClearTransientState();
                await LoadSessionBaselineAsync(port, cancellationToken);
                PublishStatus(true);

                using var connection = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
                var host = ReceiveSocketAsync(
                    new Uri($"ws://127.0.0.1:{port}/api/events.host"),
                    connection.Token);
                var mux = ReceiveSocketAsync(
                    new Uri($"ws://127.0.0.1:{port}/api/events.mux"),
                    connection.Token);
                await Task.WhenAny(host, mux);
                connection.Cancel();

                try
                {
                    await Task.WhenAll(host, mux);
                }
                catch (OperationCanceledException)
                {
                }
            }
            catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
            {
                return;
            }
            catch (Exception error)
            {
                _log.Warn("DSH event connection failed: " + error.Message);
            }

            try
            {
                await Task.Delay(delay, cancellationToken);
            }
            catch (OperationCanceledException)
            {
                return;
            }
        }
    }

    private async Task ReceiveSocketAsync(Uri uri, CancellationToken cancellationToken)
    {
        using var socket = new ClientWebSocket();
        socket.Options.SetRequestHeader("Origin", $"http://127.0.0.1:{uri.Port}");
        await socket.ConnectAsync(uri, cancellationToken);
        _log.Info("Connected to " + uri.AbsolutePath);
        SocketConnected?.Invoke(this, uri);

        var buffer = new byte[64 * 1024];
        while (socket.State == WebSocketState.Open && !cancellationToken.IsCancellationRequested)
        {
            using var idle = new CancellationTokenSource(TimeSpan.FromSeconds(90));
            using var receive = CancellationTokenSource.CreateLinkedTokenSource(
                cancellationToken,
                idle.Token);
            using var frame = new MemoryStream();
            WebSocketReceiveResult result;
            do
            {
                result = await socket.ReceiveAsync(buffer, receive.Token);
                if (result.MessageType == WebSocketMessageType.Close)
                {
                    return;
                }

                if (result.MessageType == WebSocketMessageType.Text)
                {
                    frame.Write(buffer, 0, result.Count);
                }
            }
            while (!result.EndOfMessage);

            if (frame.Length > 0)
            {
                ProcessFrame(Encoding.UTF8.GetString(frame.GetBuffer(), 0, checked((int)frame.Length)));
            }
        }
    }

    private void ProcessFrame(string json)
    {
        try
        {
            using var document = JsonDocument.Parse(json);
            var root = document.RootElement;
            if (!root.TryGetProperty("method", out var methodNode) ||
                !root.TryGetProperty("payload", out var payload))
            {
                return;
            }

            var method = methodNode.GetString();
            var sessionId = GetString(payload, "sessionId") ?? "unknown";
            lock (_sync)
            {
                switch (method)
                {
                    case "host/session-status":
                        SetMembership(_runningSessions, sessionId, GetBoolean(payload, "running"));
                        break;
                    case "question/requested":
                    {
                        var key = "question:" + sessionId;
                        _pendingQuestions.Add(key);
                        NotifyQuestion(key, payload);
                        break;
                    }
                    case "question/resolved":
                        RemoveSessionKeys(_pendingQuestions, "question:" + sessionId);
                        _notifiedAttention.Remove("question:" + sessionId);
                        break;
                    case "approval/requested":
                    {
                        var approvalId = GetString(payload, "approvalId") ?? sessionId;
                        var key = "approval:" + approvalId;
                        _pendingApprovals.Add(key);
                        NotifyApproval(key, payload);
                        break;
                    }
                    case "approval/resolved":
                    {
                        var approvalId = GetString(payload, "approvalId") ?? sessionId;
                        _pendingApprovals.Remove("approval:" + approvalId);
                        _notifiedAttention.Remove("approval:" + approvalId);
                        break;
                    }
                    case "session/event":
                        ProcessSessionEvent(sessionId, payload);
                        break;
                    case "session/jobs":
                        ProcessJobs(sessionId, payload);
                        break;
                    case "session/projection":
                        if (GetString(payload, "key") == "title")
                        {
                            var title = GetString(payload, "value");
                            if (!string.IsNullOrWhiteSpace(title)) _titles[sessionId] = title;
                        }
                        break;
                }
            }

            PublishStatus(true);
        }
        catch (JsonException error)
        {
            _log.Warn("Ignored an invalid DSH event frame: " + error.Message);
        }
    }

    private void ProcessSessionEvent(string sessionId, JsonElement payload)
    {
        if (!payload.TryGetProperty("event", out var eventNode))
        {
            return;
        }

        var eventType = GetString(eventNode, "type");
        if (eventType == "turn/start")
        {
            _runningSessions.Add(sessionId);
            return;
        }

        if (eventType != "turn/end")
        {
            return;
        }

        _runningSessions.Remove(sessionId);
        var title = _titles.TryGetValue(sessionId, out var known)
            ? known
            : "DSH conversation";
        var kind = eventNode.TryGetProperty("data", out var data) &&
                   data.TryGetProperty("reason", out var reason)
            ? GetString(reason, "kind")
            : null;
        var message = kind switch
        {
            "completed" => title + " completed.",
            "failed" => title + " failed.",
            "cancelled" => title + " was cancelled.",
            _ => title + " ended."
        };
        ScheduleCompletion(new DshNotification(
            kind == "completed" ? "Conversation completed" : "Conversation ended",
            message,
            kind == "completed" ? ToolTipIcon.Info : ToolTipIcon.Warning,
            true));
    }

    private void ProcessJobs(string sessionId, JsonElement payload)
    {
        _runningJobs.RemoveWhere(key => key.StartsWith(sessionId + ":", StringComparison.Ordinal));
        if (!payload.TryGetProperty("jobs", out var jobs) ||
            jobs.ValueKind != JsonValueKind.Array)
        {
            return;
        }

        foreach (var job in jobs.EnumerateArray())
        {
            var status = GetString(job, "status");
            if (status is "running" or "stopping")
            {
                var id = GetString(job, "id") ?? Guid.NewGuid().ToString("N");
                _runningJobs.Add(sessionId + ":" + id);
            }
        }
    }

    private void NotifyQuestion(string key, JsonElement payload)
    {
        if (!_notifiedAttention.Add(key))
        {
            return;
        }

        var message = "Open DSH to answer the pending question.";
        if (payload.TryGetProperty("questions", out var questions) &&
            questions.ValueKind == JsonValueKind.Array &&
            questions.GetArrayLength() > 0)
        {
            var first = questions[0];
            message = GetString(first, "question") ?? message;
            var header = GetString(first, "header");
            if (!string.IsNullOrWhiteSpace(header)) message = header + ": " + message;
        }

        NotificationRequested?.Invoke(this, new DshNotification(
            "DSH needs your answer",
            Limit(message, 220),
            ToolTipIcon.Warning));
    }

    private void NotifyApproval(string key, JsonElement payload)
    {
        if (!_notifiedAttention.Add(key))
        {
            return;
        }

        var tool = GetString(payload, "toolName") ?? "tool";
        var reason = GetString(payload, "reason");
        var message = string.IsNullOrWhiteSpace(reason)
            ? tool
            : tool + ": " + reason;
        NotificationRequested?.Invoke(this, new DshNotification(
            "DSH requests permission",
            Limit(message, 220),
            ToolTipIcon.Warning));
    }

    private void ScheduleCompletion(DshNotification notification)
    {
        _pendingCompletion = notification;
        _completionTimer?.Dispose();
        _completionTimer = new System.Threading.Timer(_ =>
        {
            DshNotification? pending;
            lock (_sync)
            {
                pending = _pendingCompletion;
                _pendingCompletion = null;
            }

            if (pending is not null)
            {
                NotificationRequested?.Invoke(this, pending);
            }
        }, null, TimeSpan.FromSeconds(3), Timeout.InfiniteTimeSpan);
    }

    private async Task LoadSessionBaselineAsync(int port, CancellationToken cancellationToken)
    {
        var rpcId = Guid.NewGuid().ToString();
        using var request = new HttpRequestMessage(
            HttpMethod.Post,
            $"http://127.0.0.1:{port}/api/session.list")
        {
            Content = JsonContent.Create(new
            {
                type = "client-request",
                rpcId,
                method = "session.list",
                payload = new { }
            })
        };
        using var response = await _http.SendAsync(request, cancellationToken);
        response.EnsureSuccessStatusCode();
        using var document = JsonDocument.Parse(await response.Content.ReadAsStringAsync(cancellationToken));
        var root = document.RootElement;
        if (!root.TryGetProperty("result", out var result) ||
            !result.TryGetProperty("ok", out var ok) ||
            !ok.GetBoolean() ||
            !result.TryGetProperty("value", out var value) ||
            !value.TryGetProperty("items", out var items))
        {
            return;
        }

        lock (_sync)
        {
            _runningSessions.Clear();
            foreach (var item in items.EnumerateArray())
            {
                var id = GetString(item, "sessionId");
                if (id is null)
                {
                    continue;
                }

                if (GetBoolean(item, "running")) _runningSessions.Add(id);
                if (item.TryGetProperty("projections", out var projections) &&
                    projections.TryGetProperty("values", out var values) &&
                    values.TryGetProperty("title", out var title) &&
                    title.ValueKind == JsonValueKind.String)
                {
                    _titles[id] = title.GetString()!;
                }
            }
        }
    }

    private async Task<bool> IsReadyAsync(int port, CancellationToken cancellationToken)
    {
        try
        {
            var rpcId = Guid.NewGuid().ToString();
            using var request = new HttpRequestMessage(
                HttpMethod.Post,
                $"http://127.0.0.1:{port}/api/session.list")
            {
                Content = JsonContent.Create(new
                {
                    type = "client-request",
                    rpcId,
                    method = "session.list",
                    payload = new { }
                })
            };
            using var response = await _http.SendAsync(request, cancellationToken);
            if (!response.IsSuccessStatusCode)
            {
                return false;
            }

            using var document = JsonDocument.Parse(
                await response.Content.ReadAsStringAsync(cancellationToken));
            return document.RootElement.TryGetProperty("result", out var result) &&
                   result.TryGetProperty("ok", out var ok) &&
                   ok.ValueKind == JsonValueKind.True;
        }
        catch (HttpRequestException)
        {
            return false;
        }
        catch (JsonException)
        {
            return false;
        }
        catch (TaskCanceledException) when (!cancellationToken.IsCancellationRequested)
        {
            return false;
        }
    }

    private void PublishStatus(bool serviceReady)
    {
        DshStatusSnapshot snapshot;
        lock (_sync)
        {
            if (!serviceReady)
            {
                snapshot = DshStatusSnapshot.Stopped;
            }
            else
            {
                var attention = _pendingQuestions.Count + _pendingApprovals.Count;
                var running = _runningSessions.Count + _runningJobs.Count;
                var state = attention > 0
                    ? DshActivityState.Attention
                    : running > 0
                        ? DshActivityState.Busy
                        : DshActivityState.Idle;
                var summary = attention > 0
                    ? $"DSH needs attention: {_pendingQuestions.Count} question(s), {_pendingApprovals.Count} approval(s)"
                    : running > 0
                        ? $"DSH is busy: {running} active session/job(s)"
                        : "DSH is running and idle";
                snapshot = new DshStatusSnapshot(
                    state,
                    running,
                    _pendingQuestions.Count,
                    _pendingApprovals.Count,
                    summary);
            }
        }

        StatusChanged?.Invoke(this, snapshot);
    }

    private void ClearTransientState()
    {
        lock (_sync)
        {
            _runningSessions.Clear();
            _runningJobs.Clear();
            _pendingQuestions.Clear();
            _pendingApprovals.Clear();
            _titles.Clear();
        }
    }

    private static void SetMembership(HashSet<string> set, string key, bool value)
    {
        if (value) set.Add(key);
        else set.Remove(key);
    }

    private static void RemoveSessionKeys(HashSet<string> set, string prefix) =>
        set.RemoveWhere(key => key.StartsWith(prefix, StringComparison.Ordinal));

    private static string? GetString(JsonElement element, string name) =>
        element.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.String
            ? value.GetString()
            : null;

    private static bool GetBoolean(JsonElement element, string name) =>
        element.TryGetProperty(name, out var value) &&
        value.ValueKind is JsonValueKind.True or JsonValueKind.False &&
        value.GetBoolean();

    private static string Limit(string value, int length) =>
        value.Length <= length ? value : value[..(length - 1)] + "\u2026";

    public async ValueTask DisposeAsync()
    {
        var loop = _loop;
        Stop();
        if (loop is not null)
        {
            try { await loop; } catch (OperationCanceledException) { }
        }

        _completionTimer?.Dispose();
        _http.Dispose();
    }
}
